pandas/_libs/index.pyx

```diff
@@ -544,6 +544,10 @@ cpdef convert_scalar(ndarray arr, object value):
             pass
         elif isinstance(value, timedelta):
             return Timedelta(value).value
+        elif util.is_datetime64_object(value):
+            # exclude np.datetime64("NaT") which would otherwise be picked up
+            #  by the `value != value check below
+            pass
```

Comment:
really everything from 531 down to 555 should be ripped out and this handled by the DatetimeArray/TimedeltaArray `__setitem__` implementation